### DISCUSSION TITLE
LDAP federation w/ local-only userAttributes

### DISCUSSION THREAD

**Discussion by robrez** [keycloak/keycloak#30257] [Q&A]
I need to augment the local KC user with some additional attributes while using LDAP federation.  I do not want KC to attempt to sync these attributes back to AD.  How can this be achieved?

**Comment by roamerz** [comment]
hardcoded-attribute-mapper?

Under Federation > Mappers

**Reply by robrez** [reply]
I could have explained better

> I need to augment the local KC user with some additional attributes

In the realm settings, under user profile, I've added a custom attribute. Let's call it "favoriteColor". We can now key-in a value on the user profile page. "purple".  When I click save, we must not save this back to AD.. but we must save purple in the local KC database.

A hardcoded-attribute-mapper will just save a hard-coded value into the local KC database

Here are the problems with the options available in the LDAP federation config

| Edit Mode | Problem |
|-------------|-----------|
| READ_ONLY | NO user attributes can be synced back to AD, nothing can be saved |
| WRITABLE | All user attributes attempt to sync back to AD |
| UNSYNCED | Everything is saved locally, No user attributes can make it back to AD |

It seems like we need a primitive, that does not exist, for "edit mode" on an attribute-by-attribute basis.
